common/src/main/java/com/microsoft/identity/common/internal/controllers/NativeAuthMsalController.kt

```diff
@@ -508,6 +524,303 @@ class NativeAuthMsalController : BaseNativeAuthController() {
+                is ResetPasswordSubmitApiResult.PasswordInvalid -> {
+                    ResetPasswordCommandResult.PasswordNotAccepted(
+                        error = resetPasswordSubmitApiResult.error,
+                        errorDescription = resetPasswordSubmitApiResult.errorDescription
+                    )
+                }
+
+                is ResetPasswordSubmitApiResult.ExpiredToken,
+                is ResetPasswordSubmitApiResult.UnknownError -> {
+                    Logger.warn(
+                        TAG,
+                        "Unexpected result: $resetPasswordSubmitApiResult"
+                    )
+                    resetPasswordSubmitApiResult as ApiErrorResult
+                    INativeAuthCommandResult.UnknownError(
+                        error = resetPasswordSubmitApiResult.error,
+                        errorDescription = resetPasswordSubmitApiResult.errorDescription,
+                        details = resetPasswordSubmitApiResult.details
+                    )
+                }
+            }
+        } catch (e: Exception) {
+            Logger.error(TAG, "Exception thrown in resetPasswordSubmitNewPassword", e)
+            throw e
+        }
+    }
+
+    private fun resetPasswordPollCompletion(
+        oAuth2Strategy: NativeAuthOAuth2Strategy,
+        passwordResetToken: String,
+        pollIntervalInSeconds: Int?
+    ): ResetPasswordSubmitNewPasswordCommandResult {
+        fun pollCompletionTimedOut(startTime: Long): Boolean {
+            val currentTime = System.currentTimeMillis()
+            return currentTime - startTime > ResetPasswordSubmitNewPasswordCommand.POLL_COMPLETION_TIMEOUT_IN_MILISECONDS
+        }
+
+        fun pollIntervalIsAppropriate(pollIntervalInSeconds: Int?): Boolean {
+            return pollIntervalInSeconds != null && pollIntervalInSeconds <= 15 && pollIntervalInSeconds >= 1
+        }
+
+        val methodTag = "$TAG:resetPasswordPollCompletion"
+
+        LogSession.logMethodCall(TAG, "${TAG}.resetPasswordPollCompletion")
+
+        try {
+            val pollWaitInterval: Int = if (!pollIntervalIsAppropriate(pollIntervalInSeconds)) {
+                ResetPasswordSubmitNewPasswordCommand.DEFAULT_POLL_COMPLETION_INTERVAL_IN_MILISECONDS
+            } else {
+                pollIntervalInSeconds!! * 1000
+            }
+
+            var pollCompletionApiResult = performResetPasswordPollCompletionCall(
+                oAuth2Strategy = oAuth2Strategy,
+                passwordResetToken = passwordResetToken
+            )
+
+            val startTime = System.currentTimeMillis()
+
+            while (pollCompletionApiResult is ResetPasswordPollCompletionApiResult.InProgress) {
+                // TODO: This will use coroutines, most likely shouldn't use thread sleep here
+                ThreadUtils.sleepSafely(
+                    pollWaitInterval,
+                    methodTag,
+                    "Waiting between reset password polls"
+                )
```

Comment:
We should address this, this suggestion was originally brought up by @SammyO

Reply:
I have removed the TODO as the method here wasn't a coroutine. I will consult with Sammy when she's back.